setup/gcp/installer/forseti_installer.py

```diff
@@ -0,0 +1,291 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Forseti Installer"""
+
+from __future__ import print_function
+from abc import ABCMeta
+from abc import abstractmethod
+
+import sys
+
+from util.utils import (
+    print_banner, get_forseti_version, format_service_acct_id,
+    infer_version)
+from util.constants import (
+    FORSETI_CONF_PATH, DEPLOYMENT_TEMPLATE_OUTPUT_PATH,
+    MESSAGE_DEPLOYMENT_HAD_ISSUES, MESSAGE_DEPLOYMENT_TEMPLATE_LOCATION,
+    MESSAGE_VIEW_DEPLOYMENT_DETAILS, MESSAGE_FORSETI_CONFIGURATION_GENERATED,
+    MESSAGE_FORSETI_CONFIGURATION_GENERATED_DRY_RUN, DEFAULT_BUCKET_FMT_V2,
+    MESSAGE_FORSETI_BRANCH_DEPLOYED, MAXIMUM_LOOP_COUNT)
+from util.gcloud import (
+    create_reuse_service_acct, check_billing_enabled, lookup_organization,
+    get_gcloud_info, verify_gcloud_information, create_deployment,
+    check_vm_init_status, get_vm_instance_info)
+from util.files import (
+    copy_file_to_destination, generate_deployment_templates,
+    generate_forseti_conf)
+from configs.config import Config
+
+
+class ForsetiInstaller:
+    """Forseti installer base class (abstract)"""
+    __metaclass__ = ABCMeta
+
+    # Class variables initialization
+    branch = None
+    project_id = None
+    organization_id = None
+    gcp_service_account = None
+    user_can_grant_roles = True
+    config = Config()
+
+    @abstractmethod
+    def __init__(self):
+        """Initialize."""
+        pass
+
+    def run_setup(self):
+        """Run the setup steps"""
+        print_banner('Installing Forseti {} v{}'.format(
+            self.config.template_type, get_forseti_version()))
+
+        # Preflight checks
+        self.preflight_checks()
+
+        # Deployment
+        bucket_name = self.generate_bucket_name()
+        conf_file_path = self.generate_forseti_conf()
+        deployment_tpl_path = self.generate_deployment_templates()
+
+        deploy_success, deployment_name = self.deploy(deployment_tpl_path,
+                                                      conf_file_path,
+                                                      bucket_name)
+
+        # After deployment
+        self.post_install_instructions(deploy_success,
+                                       deployment_name,
+                                       deployment_tpl_path,
+                                       conf_file_path,
+                                       bucket_name)
+
+    def preflight_checks(self):
+        """Pre-flight checks"""
+        self.check_run_properties()
+        self.branch = infer_version(self.config.advanced_mode)
+        self.project_id, authed_user, is_devshell = get_gcloud_info()
```

Comment:
`is_devshell` should rather be `is_cloud_shell`

Reply:
Updated.